rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,228 @@
+
+- ⚠ We can mistakenly debug an issue against the undeployed version of the code.
+- ⚠ We may struggle to isolate a breaking change, if many were deployed together.
+
+## Proposal
+
+Make deployments fully automatic for the apps listed [here](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/hieradata_aws/common.yaml#L145). This has already begun, as part of [a successful trial of Continuous Deployment](https://docs.google.com/document/d/18y2fLzqbHyTliThxuiyxKznDw4TZmPiQZAk-3rkgpBg/edit) and [a subsequent extension to include 5 more apps](https://docs.google.com/document/d/1YhgjJjDRB57-IlgPNAAZzFTI3jIUucGJKokti_lQiRQ/edit).
+
+> **It is still possible to deploy an app manually.** The new pipeline is just an automation of the manual deployment steps. Rollback is unchanged, and freezing deployments is supported [[1](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/modules/govuk_jenkins/templates/jobs/deploy_app_downstream.yaml.erb#L22)].
+
+We will continue to use the new development and deployment pipeline created for the trials. Much of the functionality is defined in [a new `Deploy_App_Downstream` Jenkins job](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/modules/govuk_jenkins/templates/jobs/deploy_app_downstream.yaml.erb), which works like this:
+
+1. **Integration**
+
+   1. Deployment runs.
+   1. [**new**] Wait for the Smoke tests for the app to pass [[1](https://github.com/alphagov/govuk-puppet/pull/10387)].
+   1. [**new**] Remotely trigger the same deployment in Staging.
+
+1. **Staging**
+
+   1. [**new**] Repeat as for Integration, but to Production.
+
+We propose to apply this automation across our apps. Given the scale of the change, we need to think carefully about how we can make it consistently safe and secure, before we enable CD for an app.
+
+### Safety concerns
+
+The person who merges a PR must be confident that it is safe to do so. As part of the trial of CD, [we added PR templates to each affected app](https://github.com/alphagov/smart-answers/pull/4758), to explicitly warn developers about the automatic deployment.
+
+However, we also want to ensure that a faulty PR merged in haste can be caught before it impacts the live site. There are a number of checks we can perform to help ensure this, without aiming for perfection.
+
+> **We won't check alerts.** At the time of writing, we have no evidence that checking Icinga alerts has any advantage over the checks discussed below. We can investigate such a check if we find evidence [[1](https://github.com/alphagov/govuk-puppet/pull/10597)].
+
+> **We will only consider _live_ code.** Interactions due to manual and scheduled tasks can be checked manually, using a branch deployment. In the worst case, we should have ample time to fix any faults that develop.
```

Comment:
Haha, yes "live" means "can be triggered at any time" - something like a "live wire". Do you think that needs clarifying?

One thing I've struggled on with this RFC is to contain the scope of these safety requirements: they should be just enough for us to switch from manual to automatic. This naturally excludes most rake tasks, because the risk to our users of them being broken is unchanged. So, while I definitely think we should test most of our rake tasks, so they work when they're needed, I'm seeing it as out-of-scope for the RFC.

Scheduled tasks are similar, if a little harder to argue for: the nature of them means it's unlikely we're currently doing any kind of manual testing with them (i.e. waiting for them to run), so again switching from manual to automatic should have little practical effect on the quality of them code that's shipped to production. I've managed to reassure myself that: anything on a schedule can always be done "later".

Does that make sense? Do you think others would benefit from this reasoning in the main text?

Reply:
Hmm...not sure I follow.

I think we both agree that the focus on "live" code is justified, because it's the body of code that will be most affected by enabling CD. We want to make sure the defect rate is the same or lower for this code.

For scheduled and manual tasks, we don't expect CD to change how many defects make it to production. We should obviously have tests for these things, but it's hard to argue for here, and it presents a greater barrier to entry.

> On the other hand, that same argument could apply equally to any functionality that we're not yet testing.

This is the bit I'm not sure about. 

More generally, do you think there's anything that should change in the main text?

Reply:
I'd suggest:

`**We will not require additional tests for batch and scheduled tasks** Where we currently test these manually, we will continue to do so. We do not expect continuous deployment to lead to any increase in defects reaching production for these jobs.`

(This avoids the ambiguity about "live code" and the questionable claim about worst-case bugs in batch jobs.)

Reply:
Thanks! I've reworded that slightly in https://github.com/alphagov/govuk-rfcs/pull/128/commits/2e5a53e42c7bccc679cc9577eea6bf60834aaaf2 - I think we may as well avoid any reference to how this code may (or may not) be tested, since it's out-of-scope. I'll resolve this, but do comment if you think there's more to discuss.